src/main/java/com/google/devtools/build/lib/query2/cquery/GraphOutputFormatterCallback.java

```diff
@@ -0,0 +1,115 @@
+import java.io.OutputStream;
+import java.util.Comparator;
+
+/** cquery output formatter that prints the result as factored graph in AT&amp;T GraphViz format. */
+class GraphOutputFormatterCallback extends CqueryThreadsafeCallback {
+  @Override
+  public String getName() {
+    return "graph";
+  }
+
+  /** Interface for finding a configured target's direct dependencies. */
+  @FunctionalInterface
+  public interface DepsRetriever {
+    Iterable<ConfiguredTarget> getDirectDeps(ConfiguredTarget taget) throws InterruptedException;
+  }
+
+  private final DepsRetriever depsRetriever;
+
+  private final GraphOutputWriter.NodeReader<ConfiguredTarget> nodeReader =
+      new NodeReader<ConfiguredTarget>() {
+
+        private final Comparator<ConfiguredTarget> configuredTargetOrdering =
+            (ct1, ct2) -> {
+              // Order graph output first by target label, then by configuration hash.
+              Label label1 = ct1.getLabel();
+              Label label2 = ct2.getLabel();
+              return label1.equals(label2)
+                  ? ct1.getConfigurationChecksum().compareTo(ct2.getConfigurationChecksum())
+                  : label1.compareTo(label2);
+            };
+
+        @Override
+        public String getLabel(Node<ConfiguredTarget> node) {
+          // Node payloads are ConfiguredTargets. Output node labels are target labels + config
+          // hashes.
+          ConfiguredTarget ct = node.getLabel();
+          return String.format(
+              "%s (%s)",
+              ct.getLabel(),
+              shortId(skyframeExecutor.getConfiguration(eventHandler, ct.getConfigurationKey())));
+        }
+
+        @Override
+        public Comparator<ConfiguredTarget> comparator() {
+          return configuredTargetOrdering;
+        }
+      };
+
+  GraphOutputFormatterCallback(
+      ExtendedEventHandler eventHandler,
+      CqueryOptions options,
+      OutputStream out,
+      SkyframeExecutor skyframeExecutor,
+      TargetAccessor<ConfiguredTarget> accessor,
+      DepsRetriever depsRetriever) {
+    super(eventHandler, options, out, skyframeExecutor, accessor);
+    this.depsRetriever = depsRetriever;
+  }
+
+  @Override
+  public void processOutput(Iterable<ConfiguredTarget> partialResult) throws InterruptedException {
+    // Transform the cquery-backed graph into a Digraph to make it suitable for GraphOutputWriter.
+    // Note that this involves an extra iteration over the entire query result subgraph. We could
+    // conceptually merge transformation and output writing into the same iteration if needed.
+    Digraph<ConfiguredTarget> graph = new Digraph<>();
+    ImmutableSet<ConfiguredTarget> allNodes = ImmutableSet.copyOf(partialResult);
+    for (ConfiguredTarget configuredTarget : partialResult) {
+      Node<ConfiguredTarget> node = graph.createNode(configuredTarget);
+      for (ConfiguredTarget dep : depsRetriever.getDirectDeps(configuredTarget)) {
+        if (allNodes.contains(dep)) {
```

Comment:
Assuming this is for a situation like --noimplicit_deps or the like?

Reply:
The reasoning here is that the query output may only contain a subset of all deps.

So if A depends on B, C, and D and some magical query expression filtered the results down to just A and C, `depsRetriever.getDirectDeps` returns all of A's deps (B, C, and D). We only want to include the ones that are part of th query result (any target in `partialResult`).